internal/planner/consumer.go

```diff
@@ -0,0 +1,317 @@
+	}
+
+	completedAt := job.GetCompletedAt()
+	if completedAt.IsZero() && action == "completed" {
+		return githubWebhookJob{}, fmt.Errorf("missing completed_at in completed event webhook")
+	}
+
+	id := strconv.FormatInt(job.GetID(), 10)
+	if id == "0" {
+		return githubWebhookJob{}, fmt.Errorf("missing job id in event webhook")
+	}
+
+	var rawJson map[string]interface{}
+	err = json.Unmarshal(body, &rawJson)
+	if err != nil {
+		return githubWebhookJob{}, fmt.Errorf("invalid github webhook: %v", err)
+	}
+
+	dbJob := &database.Job{
+		ID:          id,
+		Platform:    platform,
+		Labels:      job.Labels,
+		CreatedAt:   createdAt.Time,
+		StartedAt:   parseOptionalGitHubTime(job.StartedAt),
+		CompletedAt: parseOptionalGitHubTime(job.CompletedAt),
+		Raw:         map[string]interface{}{action: rawJson},
+	}
+
+	repo := jobEvent.GetRepo().GetFullName()
+
+	if repo == "" {
+		return githubWebhookJob{}, fmt.Errorf("missing repo full name in webhook payload")
+	}
+
+	return githubWebhookJob{
+		id:     id,
+		repo:   jobEvent.GetRepo().GetFullName(),
+		labels: job.Labels,
+		action: action,
+
+		payload: jobEvent,
+
+		job: dbJob,
+	}, nil
+}
+
+// Start begins consuming messages from the AMQP queue and processing them.
+func (c *JobConsumer) Start(ctx context.Context) error {
+	logger.InfoContext(ctx, "start consume workflow jobs from queue")
+	backoff := time.Second
+	for {
+		_ = c.pullMessage(ctx)
+		select {
+		case <-time.After(backoff):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+		backoff = min(5*time.Minute, backoff*2)
+	}
+}
+
+// pullMessage receive one message from the AMQP queue and process it.
+// pullMessage can return errors, but it's mostly for testing and can be ignored,
+// as all errors are handled internally before return.
+func (c *JobConsumer) pullMessage(ctx context.Context) error {
+	msg, err := c.consumer.Pull(ctx)
+	if err != nil {
+		logger.ErrorContext(ctx, "cannot receive message from amqp, retrying", "error", err)
+		c.metrics.ObserveWebhookError(ctx, platform)
+		return err
+	}
+	headers := make(map[string]string)
+	for h, v := range msg.Headers {
+		if strVal, ok := v.(string); ok {
+			headers[h] = strVal
+		}
+	}
+	ctx = otel.GetTextMapPropagator().Extract(ctx, propagation.MapCarrier(headers))
+	ctx, span := trace.Start(ctx, "consume webhook")
+	defer span.End()
+
+	job, err := parseGithubWebhookPayload(ctx, msg.Headers, msg.Body)
+	if err != nil {
+		span.RecordError(err)
+		c.metrics.ObserveWebhookError(ctx, platform)
+		c.discardMessage(ctx, &msg)
+		return err
+	}
+
+	if !slices.Contains([]string{"queued", "in_progress", "completed"}, job.action) {
+		logger.InfoContext(ctx, "ignore other action types for GitHub webhook job", "action", job.action)
+		c.metrics.ObserveDiscardedWebhook(ctx, platform)
+		c.discardMessage(ctx, &msg)
+		return nil
+	}
+
+	isSelfHosted := false
+	for _, label := range job.labels {
+		if strings.Contains(label, "self-hosted") {
+			isSelfHosted = true
+			break
+		}
+	}
```

Comment:
can we put this logic into another function, e.g. `isSelfHostedJob(label: string) bool`? I think this is crucial logic which should be handled by a dedicated function. 

This would improve readability and also obey Single Responsibility Principle on function level